ls1intum/Artemis
# `Exam mode`: Improve usability of attendance check

### Checklist
#### General



#### Client

### Motivation and Context


Exam attendance check feature is not easy to use and the UI is difficult to understand.

### Description


- Instead of empty list in Verify Attendance, show a green message that no issues appeared
- If there is an issue, also show how many issues appear
- Merge columns Planned Room  Planned Seat  Actual Room  Actual Seat  into two columns Room and Seat
- New column with a status of the attendance check (Not attended, Attendance Check failed , Attendance Check passed)
- only show status column after exam has finished

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Course with exam

1. Log in to Artemis
2. Navigate to Course Administration
3. Register students to exam 
4. Upload the pdf that contains student images, check if images are saved correctly.
5. Generate Student Exams (check if you can fetch all student exams and view specific student exam) 
6. Login as student 
7. Participate in exam and submit the exam
8. Log as instructor back and verify attendance (You should see students that started the exam but did not sign in a table. )
9. If all students where checked, you should see a message that all students were checked successfully. 
10. Also in the students table you should see a new column called status after the exam has ended (statuses: Not attended, Attendance Check failed , Attendance Check passed)

Example files: 
[studentsWithImages.pdf](https://github.com/ls1intum/Artemis/files/10910361/studentsWithImages.pdf)
[studentsArtemis (1).csv](https://github.com/ls1intum/Artemis/files/10910367/studentsArtemis.1.csv)

Note:In Order for students to sign, you need to have IPad and the exam-attendance-checker app downloaded. But if you want you can change/update exam-user information by making a request to the following endpoint (use postman): 

``` 
/**
     * POST /courses/:courseId/exams/:examId/exam-users : Update the exam user with the given DTO info
     *
     * @param examUserDTO   the dto containing exam user info
     * @param signatureFile the signature of the student
     * @param courseId      the id of the course
     * @param examId        the id of the exam
     * @return saved examUser ResponseEntity with status 200 (OK) or with status 404 (Not Found)
     */
    @PostMapping("courses/{courseId}/exams/{examId}/exam-users")
    @PreAuthorize("hasRole('INSTRUCTOR')")
    public ResponseEntity<ExamUser> updateExamUser(@RequestPart ExamUserDTO examUserDTO, @RequestPart(value = "file", required = false) MultipartFile signatureFile,
            @PathVariable Long courseId, @PathVariable Long examId) {

....................
}
 ```
Here you can see an postman request example: 

![exampleOfPostman](https://user-images.githubusercontent.com/26312757/219153178-b1f33a07-8f81-4874-9185-72112ddc5bcc.png)
examUserDTO object is as follow:

```
{
 "login": "student",
"didCheckImage":true,
"didCheckLogin":true,
"didCheckName":true,
"didCheckRegistrationNumber":true
}
```
Note: login field is important as its used as identifier for students

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests


### Screenshots



Students page before:

![Screenshot 2023-04-28 at 09 53 50](https://user-images.githubusercontent.com/26312757/235094928-7e8c3973-5ef3-4a80-9b64-3cf34c71dbe9.jpg)

Students page after:
![Screenshot 2023-04-28 at 17 34 52](https://user-images.githubusercontent.com/26312757/235197179-d25c17d3-4468-4db0-ae3e-ef7d91fe35a2.jpg)



Verify attendance check Before:
![Screenshot 2023-04-28 at 09 52 52](https://user-images.githubusercontent.com/26312757/235095045-ed5e1bb8-dba9-4d58-814b-b06b9f765149.jpg)

![Screenshot 2023-04-28 at 09 52 26](https://user-images.githubusercontent.com/26312757/235095153-b2142309-5530-4d87-b919-88f40d06ee04.jpg)


Verify attendance check After:
![Screenshot 2023-04-28 at 09 51 28](https://user-images.githubusercontent.com/26312757/235095195-c97ac37a-34c5-4b0d-889c-1c376aeb5f6d.jpg)

![Screenshot 2023-04-28 at 10 28 08](https://user-images.githubusercontent.com/26312757/235096542-3f05c367-b675-44ad-9fa5-4ae5bc172052.jpg)

Updated room/seat columns:

![Screenshot 2023-04-28 at 10 00 55](https://user-images.githubusercontent.com/26312757/236677468-6a281aca-6943-42b1-80e0-0786db2ef980.jpg)
